p4_symbolic/symbolic/values.cc

```diff
@@ -0,0 +1,312 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file is responsible for parsing values from the bmv2 json and table
+// entries.
+// It is also responsible for translating any string values to corresponding
+// bitvectors and back, for fields that have the @p4runtime_translation
+// annotation.
+
+#include "p4_symbolic/symbolic/values.h"
+
+#include <cstdint>
+#include <locale>
+#include <sstream>
+#include <vector>
+
+#include "absl/strings/match.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_format.h"
+#include "absl/strings/str_split.h"
+#include "absl/strings/string_view.h"
+#include "absl/strings/strip.h"
+#include "p4_pdpi/utils/ir.h"
+#include "p4_symbolic/symbolic/operators.h"
+#include "p4_symbolic/symbolic/symbolic.h"
+
+namespace p4_symbolic {
+namespace symbolic {
+namespace values {
+
+namespace {
+
+// Finds the minimum bit size required for representing the given value.
+unsigned int FindBitsize(uint64_t value) {
+  unsigned int bitsize = 0;
+  uint64_t pow = 1;
+  while (bitsize <= 64 && pow <= value) {
+    pow = pow * 2;
+    bitsize++;
+  }
+  return (bitsize > 1 ? bitsize : 1);  // At least 1 bit.
+}
+
+// Normalize the given value: transforms it into a binary representation
+// without any leading zeros.
+std::string NormalizeBits(std::string value) {
+  static std::unordered_map<char, std::string> hex_to_bin = {
+      {'0', "0000"}, {'1', "0001"}, {'2', "0010"}, {'3', "0011"},
+      {'4', "0100"}, {'5', "0101"}, {'6', "0110"}, {'7', "0111"},
+      {'8', "1000"}, {'9', "1001"}, {'a', "1010"}, {'b', "1011"},
+      {'c', "1100"}, {'d', "1101"}, {'e', "1110"}, {'f', "1111"}};
+
+  std::string stripped = "";
+  if (absl::StartsWith(value, "#x")) {
+    // Turn hex to binary.
+    absl::string_view stripped_value = absl::StripPrefix(value, "#x");
+    for (char c : stripped_value) {
+      absl::StrAppend(&stripped, hex_to_bin.at(c));
+    }
+  } else if (absl::StartsWith(value, "#b")) {
+    // Strip leading #b for binary strings.
+    stripped = absl::StripPrefix(value, "#b");
+  } else {
+    // No normalization requied for bools or ints.
+    return value;
+  }
+
+  // Remove leading zeros.
+  size_t first_zero_index = stripped.size() - 1;
+  for (size_t i = 0; i < stripped.size(); i++) {
+    if (stripped.at(i) == '1') {
+      first_zero_index = i;
+      break;
+    }
+  }
+  return stripped.substr(first_zero_index);
+}
+
+}  // namespace
+
```

Comment:
End of new code, below is a portion of existing code that is copied from a different file for organization purposes.